src/main/java/com/iota/iri/utils/thread/BoundedScheduledExecutorService.java

```diff
@@ -0,0 +1,931 @@
+            return delegate().getDelay(unit);
+        }
+
+        @Override
+        public int compareTo(Delayed o) {
+            return delegate().compareTo(o);
+        }
+
+        /**
+         * {@inheritDoc}
+         * <br />
+         * This method fires the {@link #onCancelTask(TaskDetails)} if the future has not been cancelled before.
+         * Afterwards it also fires the {@link #onCompleteTask(TaskDetails, Throwable)} callback if the task is not
+         * running right now (otherwise this will be fired inside the wrapped task).<br />
+         */
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            if (!delegate().isCancelled() && !delegate().isDone()) {
+                onCancelTask(taskDetails);
+            }
+
+            if (taskDetails.getScheduledForExecution().compareAndSet(true, false)) {
+                onCompleteTask(taskDetails, null);
+            }
+
+            return delegate().cancel(mayInterruptIfRunning);
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return delegate().isCancelled();
+        }
+    }
+
+    /**
+     * This method wraps the passed in task to automatically call the callbacks for its lifecycle.<br />
+     * <br />
+     * The lifecycle methods are for example used to manage the resources that are reserved for this task.<br />
+     *
+     * @param task the raw task that shall be wrapped with the resource freeing logic
+     * @param taskDetails metadata holding the relevant information of the task
+     * @param future the wrapped {@link Future} that this task belongs to - it allows us to check if the task was
+     *               cancelled from the outside
+     * @param <V> the return type of the task
+     * @return a wrapped task that cleans up its reserved resources upon completion
+     */
+    private <V> Callable<V> wrapTask(Callable<V> task, TaskDetails taskDetails, Future<V> future) {
+        onScheduleTask(taskDetails);
+
+        return () -> {
+            if (taskDetails.getScheduledForExecution().compareAndSet(true, false)) {
+                Throwable error = null;
+                try {
+                    onStartTask(taskDetails);
+                    taskDetails.getExecutionCount().incrementAndGet();
+
+                    return task.call();
+                } catch (Exception e) {
+                    error = e;
+
+                    throw e;
+                } finally {
+                    onFinishTask(taskDetails, error);
+
+                    if (taskDetails.getInterval() == null || error != null || future == null || future.isCancelled() ||
+                            delegate.isShutdown()) {
+
+                        onCompleteTask(taskDetails, error);
+                    } else {
+                        taskDetails.getScheduledForExecution().set(true);
+                    }
+                }
+            }
+
+            return null;
```

Comment:
Are you sure you handled all the cases to prevent later `NullPointerExceptions`?
If the answer is no, maybe it is better to use `Optional`?

Reply:
Unless you can explain to me how you are sure there are no `NullPointerExceptions` please do that :-)

Reply:
I don't get it - the only way to retrieve the result is by using future.get() which will already throw a CancellationException if the future was cancelled.

This return statement can only be reached "internally" but the null will never be accessible to any caller.

Reply:
Hmm
You create a `Callable` and sometimes you pass it to an executor (`delegate.invokeAny()` for example) as the task to be invoked.

According to the javadocs of `invokeAny` (or any of the methods that try to run the `Callable` tasks), a NullPointerException will be thrown...

Reply:
Yes, if the Callable itself is null but not if the callable return null.

So if I do sth like executorService.schedule...(null);